src/dict.c

```diff
@@ -872,142 +880,169 @@ void dictTwoPhaseUnlinkFree(dict *d, dictEntry *he, dictEntry **plink, int table
-    return de->v.s64;
+    return DICT_GET_VALUE(de, v.s64);
 }
 
 uint64_t dictGetUnsignedIntegerVal(const dictEntry *de) {
-    assert(entryHasValue(de));
-    return de->v.u64;
+    return DICT_GET_VALUE(de, v.u64);
 }
 
 double dictGetDoubleVal(const dictEntry *de) {
-    assert(entryHasValue(de));
-    return de->v.d;
+    return DICT_GET_VALUE(de, v.d);
 }
 
 /* Returns a mutable reference to the value as a double within the entry. */
 double *dictGetDoubleValPtr(dictEntry *de) {
-    assert(entryHasValue(de));
-    return &de->v.d;
+    return DICT_GET_VALUE_PTR(de, v.d);
 }
 
 /* Returns the 'next' field of the entry or NULL if the entry doesn't have a
  * 'next' field. */
 static dictEntry *dictGetNext(const dictEntry *de) {
     if (entryIsKey(de)) return NULL; /* there's no next */
     if (entryIsNoValue(de)) return decodeEntryNoValue(de)->next;
-    if (entryIsEmbedded(de)) return decodeEmbeddedEntry(de)->next;
-    return de->next;
+    if (entryIsEmbedded(de)) return decodeEntryEmbedded(de)->next;
+    return decodeEntryNormal(de)->next;
 }
 
 /* Returns a pointer to the 'next' field in the entry or NULL if the entry
  * doesn't have a next field. */
 static dictEntry **dictGetNextRef(dictEntry *de) {
     if (entryIsKey(de)) return NULL;
     if (entryIsNoValue(de)) return &decodeEntryNoValue(de)->next;
-    if (entryIsEmbedded(de)) return &decodeEmbeddedEntry(de)->next;
-    return &de->next;
+    if (entryIsEmbedded(de)) return &decodeEntryEmbedded(de)->next;
+    return &decodeEntryNormal(de)->next;
 }
 
 static void dictSetNext(dictEntry *de, dictEntry *next) {
-    assert(!entryIsKey(de));
     if (entryIsNoValue(de)) {
         decodeEntryNoValue(de)->next = next;
     } else if (entryIsEmbedded(de)) {
-        decodeEmbeddedEntry(de)->next = next;
+        decodeEntryEmbedded(de)->next = next;
     } else {
-        de->next = next;
+        assert(entryIsNormal(de));
+        decodeEntryNormal(de)->next = next;
     }
 }
 
 /* Returns the memory usage in bytes of the dict, excluding the size of the keys
  * and values. */
 size_t dictMemUsage(const dict *d) {
-    return dictSize(d) * sizeof(dictEntry) + dictBuckets(d) * sizeof(dictEntry *);
+    return dictSize(d) * sizeof(dictEntryNormal) + dictBuckets(d) * sizeof(dictEntry *);
 }
 
 /* Returns the memory usage in bytes of dictEntry based on the type. if `de` is NULL, return the size of
  * regular dict entry else return based on the type. */
 size_t dictEntryMemUsage(dictEntry *de) {
     if (de == NULL || entryIsNormal(de))
-        return sizeof(dictEntry);
+        return sizeof(dictEntryNormal);
     else if (entryIsKey(de))
         return 0;
     else if (entryIsNoValue(de))
         return sizeof(dictEntryNoValue);
     else if (entryIsEmbedded(de))
-        return zmalloc_size(decodeEmbeddedEntry(de));
+        return zmalloc_size(decodeEntryEmbedded(de));
     else
-        assert("Entry type not supported");
+        assert(!"Entry type not supported");
```

Comment:
```suggestion
        panic("Entry type not supported");
```
Isn't this just the wrong function?

Reply:
yeah, make sense. `Panic` is much better.